### PR TITLE
Layout paragraph using the appropriate width based on its textWidthBasis

### LINKED ISSUE #62550: With RTL, the RichText content moves but the widget does not
See repro:

```dart
Container(
    padding: EdgeInsets.all(matGrid * 2),
    child: Column(
        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text('LTR Text Span (Default TextWidthBasis):',
              style: Theme.of(context).textTheme.caption),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                TextSpan(children: [
                  TextSpan(text: 'this is a line of text with words'),
                ]),
              )),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                TextSpan(children: [
                  TextSpan(
                      text:
                          'this is a line of text with words followed by areallylongwordthatneedstobewrapped'),
                ]),
              )),
          Text('RTL Text Span (Default TextWidthBasis):',
              style: Theme.of(context).textTheme.caption),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(text: 'this is a line of text with words'),
                  ]),
                  textDirection: TextDirection.rtl)),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(
                        text:
                            'this is a line of text with words followed by areallylongwordthatneedstobewrapped'),
                  ]),
                  textDirection: TextDirection.rtl)),
          Text('LTR Text Span (TextWidthBasis.longestLine):',
              style: Theme.of(context).textTheme.caption),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(text: 'this is a line of text with words'),
                  ]),
                  textWidthBasis: TextWidthBasis.longestLine)),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(
                        text:
                            'this is a line of text with words followed by areallylongwordthatneedstobewrapped'),
                  ]),
                  textWidthBasis: TextWidthBasis.longestLine)),
          Text('RTL Text Span (TextWidthBasis.longestLine):',
              style: Theme.of(context).textTheme.caption),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(text: 'this is a line of text with words'),
                  ]),
                  textWidthBasis: TextWidthBasis.longestLine,
                  textDirection: TextDirection.rtl)),
          Container(
              decoration: BoxDecoration(border: Border.all(width: 1)),
              child: Text.rich(
                  TextSpan(children: [
                    TextSpan(
                        text:
                            'this is a line of text with words followed by areallylongwordthatneedstobewrapped '),
                  ]),
                  textWidthBasis: TextWidthBasis.longestLine,
                  textDirection: TextDirection.rtl)),
        ]));
```

Produces the following screen (screen does not map to code directly. Customer reported two issues in a single repro and I edited the code):

image

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

When a `TextPainter.textWidthBasis` is `longestLine`, the `TextPainter` should layout its paraph using the width of its longest line from the 1st layout pass. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/62550

## Tests

I added the following tests:

    'textWidthBasis.longestLine confines the width of the paragraph '
    'when given loose constraints'

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by GaryQian — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 1]
```diff
@@ -568,7 +568,21 @@ class TextPainter {
...
+        case TextWidthBasis.longestLine:
+          // The parent widget expects the paragraph to be exactly
+          // `TextPainter.width` wide, if that value satisfies the constraints
+          // it gave to the TextPainter. So when `textWidthBasis` is longestLine,
+          // the paragraph's width needs to be as close to the width of its
+          // longest line as possible.
+          newWidth = _applyFloatingPointHack(_paragraph.longestLine);
+          break;
+        case TextWidthBasis.parent:
+          newWidth = maxIntrinsicWidth;
```
Comment (justinmc) [inline packages/flutter/lib/src/painting/text_painter.dart:582]: Does this need to be clamped like it was before?

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Thanks, I probably created this bug when I introduced TextWidthBasis.

File: packages/flutter/lib/src/painting/text_painter.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/painting/text_painter.dart:582]: it gets clamped after the `switch`.

Comment (justinmc) [reply packages/flutter/lib/src/painting/text_painter.dart:582]: Ah I see, thanks.

**Pull request merged**